### PR TITLE
Sessions

### LINKED ISSUE #117: Session support
There are two problems with the system right now.

1. A person should not have to insert their username again and again
2. If someone solves a question, they have no way of automatically proving that they solved it (someone else could have submitted on their name)

A simple/bare bones solution for this is to have something called a session:

- ask a person to register/ provide two pieces of information
    - public information = username
    - private information = password
- then whenever needed, ask them to prove their identity by providing both pieces of info
- on any api which needs to verify / access user details ask for these pairs

Asking for these pairs again and again becomes irritating very quickly. Instead what we do is:

- ask them once
- send them a random number
- later on they can supply this random number instead of their details and we will know that it's them by looking up in some data structure internally.
- this number/token exchange thing is carried out using cookies. Thus, the browser handles everything for us

Implementation
-------------------

1. login/registration/logout api
2. login sets a cookie with a random number in the cookie
3. everyone uses cookies to verify identity by looking up in some data structure (dict maybe?)
4. logout un-sets the cookie

### PR DESCRIPTION
**Pull request by rishabhKalakoti**
TODOs
- [x] basic login/ register interface and handler
- [x] logged_in function that matches cookies to check if a user is logged in
- [x] add accessibility check for handlers using logged_in
- [x] create_session function to create a new session if login successful
- [x] finalize: logout and cookie expiry

fixes #117

### CODE REVIEW

**Comment by rishabhKalakoti** [comment]
@theSage21 need some help with cookies in python, i checked a few links but can't handle them till now... plz help regarding setting/ deleting/ retriving cookies

**Comment by theSage21** [comment]
Ah yes. Cookies are a pain to work with if you don't fully understand them. There are a few things to keep in mind:

1. cookies are set via response headers
2. browser can completely ignore them :smile:  if it so chooses
3. cookies have a path. They are sent only for paths which contain that path. Default is the path that set the cookie
4. they have an age
5. they are sent only for certain domains

I suggest going through [bottle's cookies docs](https://bottlepy.org/docs/dev/tutorial.html#cookies) and reading through all the options present there. Might help you understand a little better. In fact, see the code if it's still not clear.

Finally you'll have to do something like:

```python
bottle.response.set_cookie(name, value, path='/')
```

**Comment by rishabhKalakoti** [comment]
What else can I add in this one? The ones mentioned in the issue have been handled

File: server.py [hunk 1]
```diff
@@ -150,9 +170,73 @@ def rankings():
...
+    return redirect("/dashboard")
+
+
+@app.post("/login")
+def login():
+    username = request.forms.get("username")
+    password = request.forms.get("password")
+    with shelve.open(user_db) as users:
+        if not username in users:
+            return "user does not exist."
```
Comment (theSage21) [inline server.py]: let's use an `abort` here

Comment (theSage21) [reply server.py]: `return abort(404, 'user does not exist')`

File: server.py [hunk 2]
```diff
@@ -150,9 +170,73 @@ def rankings():
...
+        if users[username].password != password:
+            return "incorrect password."
```
Comment (theSage21) [inline server.py]:
```suggestion
            return abort(403, 'incorrect user-password combo')
```

File: server.py [hunk 3]
```diff
@@ -150,9 +170,73 @@ def rankings():
...
+
+
+@app.post("/register")
+def register():
+    username = request.forms.get("username")
+    password = request.forms.get("password")
+    cpassword = request.forms.get("cpassword")
+    firstname = request.forms.get("firstname")
+    lastname = request.forms.get("lastname")
+    if password != cpassword:
```
Comment (theSage21) [inline server.py]: this validation can be done in the UI

File: server.py [hunk 4]
```diff
@@ -150,9 +170,73 @@ def rankings():
...
+            password=password, firstname=firstname, lastname=lastname
+        )
+    return createSession(username)
+
+
+@app.get("/logout")
+def logout():
+    with shelve.open(sessions_db) as sessions:
+        del sessions[request.get_cookie("s_id")]
+    response.set_cookie("s_id", "")
```
Comment (theSage21) [inline server.py]: maybe use delete_cookie here?

File: views/home.html [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+                <form name="register" method="post" action="/register">
+                    <label>Username:</label>
+                    <input type="text" name="username" required /><br />
+                    <label>First name:</label>
+                    <input type="text" name="firstname" required /><br />
+                    <label>Last name:</label>
+                    <input type="text" name="lastname" /><br />
+                    <label>Password:</label>
+                    <input type="password" name="password" required /><br />
+                    <label>Confirm Password:</label>
```
Comment (theSage21) [inline views/home.html]: let's get rid of this part? Instead just show the password in plain text when registering and the ability to toggle it (if you're feeling up to it. I'm ok with a plaintext password)

File: server.py (hunk 1 above)
Comment (rishabhKalakoti) [reply server.py]: I was thinking of sending the user back to login page and `user does not exist will be shown ` on the same page, that's why I just returned the text here and left if for later.

Comment (theSage21) [reply server.py]: ah. makes sense. Let's do that then

File: views/home.html (hunk 5 above)
Comment (rishabhKalakoti) [reply views/home.html]: I would prefer hidden instead of plaintext as it is a password field, your call.

File: server.py (hunk 1 above)
Comment (rishabhKalakoti) [reply server.py]: I will code it and then you can review

**Comment by theSage21** [comment]
We can do a single hidden field + show password button or something.
I don't like double password entries. 😂

**Comment by rishabhKalakoti** [comment]
I tried something. Let me know if the decorator and message displaying is satisfactory. I kept just one password for now.

File: server.py [hunk 6]
```diff
@@ -48,22 +49,36 @@
...
-    # read the correct output as bytes object
     with open(os.path.join(question_dir, i, "output.txt"), "rb") as fl:
         output = fl.read()
     with open(os.path.join(question_dir, i, "statement.txt"), "r") as fl:
         statement = fl.read()
     questions[i] = Question(output=output, statement=statement)
 
 
+def login_required(function):
+    def login_redirect():
```
Comment (theSage21) [inline server.py]:
```suggestion
    def login_redirect(*args, **kwargs):
```

File: server.py [hunk 7]
```diff
@@ -48,22 +49,36 @@
...
+        if not logggedIn():
+            return bottle.template("home.html", message="Login required.")
+        return function()
```
Comment (theSage21) [inline server.py]:
```suggestion
        return function(*args, **kwargs)
```

File: server.py [hunk 8]
```diff
@@ -48,22 +49,36 @@
...
+    return login_redirect
+
 @app.route("/")
 def changePath():
-    return redirect("/dashboard")
+    return bottle.redirect("/dashboard")
+
+
+@app.get("/home")
```
Comment (theSage21) [inline server.py]: this can be the `/` url?

File: server.py [hunk 9]
```diff
@@ -48,22 +49,36 @@
...
 
 
 @app.get("/dashboard")
+@login_required
 def dashboard():
-    return template("dashboard.html", contests=contests)
+    return bottle.template("dashboard.html", contests=contests)
 
 
 @app.get("/contest/<code>/<number>")
```
Comment (theSage21) [inline server.py]: maybe a login_required here?

File: server.py [hunk 10]
```diff
@@ -73,28 +88,30 @@ def contest(code, number):
...
-    return template(
-        "index.html", question_number=number, contest=code, question=statement
+    return bottle.template(
+        "question.html", question_number=number, contest=code, question=statement
     )
 
 
 @app.get("/contest/<code>")
 def contest(code):
+    if not logggedIn():
```
Comment (theSage21) [inline server.py]: we can get rid of this now? Use login_required?

File: server.py (hunk 7 above)
Comment (theSage21) [reply server.py]: otherwise routes like `/some/<variable>` won't work

**Pull request merged**